src/wrapped.jl

```diff
@@ -0,0 +1,317 @@
+using Distributions
+
+"""
+    wrapped(d0::UnivariateDistribution, l::Real, u::Real; k::Int=1, kwargs...)
+
+Construct a ``k``-wrapped distribution with support ``[l, u)``.
+
+The resulting density wraps around the interval. With the period ``\\tau = u - l``, the
+density can be expressed as two infinite series in terms of either the density ``f_0`` of
+the unwrapped distribution `d0` or in terms of its characteristic function [`cf`](@ref)
+``\\phi_0``:
+````math
+\\begin{aligned}
+f(x) &= \\frac{1}{\\tau}\\sum_{j=-\\infty}^\\infty f_0\\left(x + \\frac{j}{k}\\tau\\right)\\
+     &= \\frac{1}{\\tau}\\sum_{j=-\\infty}^\\infty
+            \\phi_0\\left(jk\\frac{2\\pi}{\\tau}\\right)
+            \\exp\\left(-ijk\\frac{2\\pi}{\\tau}x\\right)
+\\end{aligned}
+````
+If `k>1`, then the resulting distribution has ``k``-fold symmetry in the interval.
+
+# Keywords
+- `characteristic::Bool`: whether to use the characteristic function to compute the density.
+    If `false`, the density is computed using the logpdf of the unwrapped distribution.
+    If not specified, the `wrapped` method called may choose based on the parameters of
+    `l`, `u`, `tol`, and `d0`.
+- `tol`: target absolute error of `pdf`. May be used to determine whether to use the
+    characteristic function in computations or not.
+
+# Examples
+
+"""
+function wrapped(d::UnivariateDistribution, l::Real, u::Real; kwargs...)
+    return Wrapped(d, promote(l, u)...; kwargs...)
+end
+function wrapped(d::UnivariateDistribution, l::T, u::T; kwargs...) where {T<:Real}
+    return Wrapped(d, l, u; kwargs...)
+end
+
+"""
+    Wrapped
+
+Generic wrapper for a wrapped distribution.
+"""
+struct Wrapped{D<:UnivariateDistribution,S<:ValueSupport,T<:Real} <:
+       UnivariateDistribution{S}
+    unwrapped::D          # the original distribution (untruncated)
+    lower::T              # lower bound
+    upper::T              # upper bound
+    characteristic::Bool  # whether to use characteristic function
+    k::Int                # k-fold symmetry
+    function Wrapped(
+        d::UnivariateDistribution,
+        l::T,
+        u::T;
+        tol::Real=sqrt(eps(float(T))),
+        characteristic::Bool=_wrapped_use_characteristic(d, l, u, tol),
```

Comment:
This seems like an odd thing to store in a field and expose to the user. Naively, it seems like an implementation detail that an average user would not need to necessarily care about or modify. I would expect that whether the characteristic function is used would be up to methods defined for particular `D`s in `Wrapped{D}`.

Reply:
I'm not sure I understand the question. There is no one number of terms. For a given wrapped distribution `dist` with values of `params`, one or both of the series will have fewer terms. But if you change any of the `params` values or change the `dist`, then the other series might now have fewer terms. The difference can be quite extreme. For wrapped normal, for example, if `σ=1e+3`, the series using `cf` needs only 1 term, while the series using pdf needs  >2,000 terms. But for `σ=1e-3`, the series using `cf` needs >6,000 terms, while the one using `pdf` needs 1 term. So we need to choose whether to use the `pdf` series or `cf` series. Options:
- always use one or the other. This is bad for the reasons noted above
- store a field like we do here. The user can select which strategy they want to use, or a developer can overloaded `wrapped(::SomeDistribution, ...)` if they have a heuristic for choosing the strategy
- Instead of storing a field, the strategy coul be picked at every evaluation of `pdf`, `cdf`, etc. Likely wasteful
- Instead of choosing, every function evaluates  both series in parallel until one of them converges. Potentially runs into issues when `cf` is not implemented, and does twice as much works as might be necessary.

Reply:
> I'm not sure I understand the question. There is no one number of terms.

Sorry, I was referring to what you had said in the preceding message:

> this requires someone to sit down and work out bounds for the number of terms needed for each series representation.